robot/web/stream/server.go

```diff
@@ -270,28 +231,51 @@ func (srs *streamRPCServer) RemoveStream(ctx context.Context, req *streampb.Remo
 		return nil, fmt.Errorf("no stream for %q", req.Name)
 	}
 
-	srs.ss.mu.Lock()
-	defer srs.ss.mu.Unlock()
+	ss.mu.Lock()
+	defer ss.mu.Unlock()
 
-	if _, ok := srs.ss.activePeerStreams[pc][req.Name]; !ok {
+	if _, ok := ss.activePeerStreams[pc][req.Name]; !ok {
 		return nil, errors.New("stream already inactive")
 	}
 	defer func() {
-		delete(srs.ss.activePeerStreams[pc], req.Name)
+		delete(ss.activePeerStreams[pc], req.Name)
 	}()
 
 	var errs error
-	for _, sender := range srs.ss.activePeerStreams[pc][req.Name].senders {
+	for _, sender := range ss.activePeerStreams[pc][req.Name].senders {
 		errs = multierr.Combine(errs, pc.RemoveTrack(sender))
 	}
 	if errs != nil {
 		return nil, errs
 	}
-	srs.ss.activeBackgroundWorkers.Add(1)
+	ss.activeBackgroundWorkers.Add(1)
 	utils.PanicCapturingGo(func() {
-		defer srs.ss.activeBackgroundWorkers.Done()
+		defer ss.activeBackgroundWorkers.Done()
 		streamToRemove.Stop()
 	})
 
 	return &streampb.RemoveStreamResponse{}, nil
 }
+
+// Close closes the Server and waits for spun off goroutines to complete.
+func (ss *Server) Close() error {
+	ss.mu.RLock()
```

Comment:
Aren't we writing to `ss.streams[N].activePeers`? If so, why are we taking a read lock rather than a write lock?

Reply:
Yeah this is just wrong and I'm not doing to touch it. The read-lock I believe is intended for accessing `ss.streams`. And this read-lock will incidentally block out [concurrent callers to](https://github.com/viamrobotics/rdk/pull/3699/files#diff-b664e7f5f24e04be1ed075bf7d0f4209eaf9fc88f34eb7acc987ad1693857223R40-R48) `streamState.Stop`. But it won't block out concurrent callers to `Server.Stop`.

I'm leaving this as-is, but the smell bothers me to. I think we got a little too cute trying to trying to avoid adding a `streamState.Close/Shutdown` that avoids the reference counting.